custom/src/main/java/com/splunk/opentelemetry/servicename/JarServiceNameProvider.java

```diff
@@ -0,0 +1,150 @@
+ * limitations under the License.
+ */
+
+package com.splunk.opentelemetry.servicename;
+
+import static java.util.logging.Level.FINE;
+
+import com.google.auto.service.AutoService;
+import io.opentelemetry.api.common.Attributes;
+import io.opentelemetry.sdk.autoconfigure.spi.ConfigProperties;
+import io.opentelemetry.sdk.autoconfigure.spi.ResourceProvider;
+import io.opentelemetry.sdk.autoconfigure.spi.internal.ConditionalResourceProvider;
+import io.opentelemetry.sdk.resources.Resource;
+import io.opentelemetry.semconv.resource.attributes.ResourceAttributes;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+@AutoService(ResourceProvider.class)
+public class JarServiceNameProvider implements ConditionalResourceProvider {
+
+  private static final Logger logger = Logger.getLogger(ServletServiceNameProvider.class.getName());
+
+  private final Supplier<String[]> getProcessHandleArguments;
+  private final Function<String, String> getSystemProperty;
+  private final Predicate<Path> fileExists;
+
+  @SuppressWarnings("unused") // SPI
+  public JarServiceNameProvider() {
+    this(JarServiceNameProvider::getArgumentsFromProcessHandle, System::getProperty, Files::exists);
+  }
+
+  // visible for tests
+  JarServiceNameProvider(
+      Supplier<String[]> getProcessHandleArguments,
+      Function<String, String> getSystemProperty,
+      Predicate<Path> fileExists) {
+    this.getProcessHandleArguments = getProcessHandleArguments;
+    this.getSystemProperty = getSystemProperty;
+    this.fileExists = fileExists;
+  }
+
+  @Override
+  public Resource createResource(ConfigProperties config) {
+    Path jarPath = getJarPathFromProcessHandle();
+    if (jarPath == null) {
+      jarPath = getJarPathFromSunCommandLine();
+    }
+    if (jarPath == null) {
+      return Resource.empty();
+    }
+    String serviceName = getServiceName(jarPath);
+    logger.log(FINE, "Auto-detected service name from the jar file name: {0}", serviceName);
+    return Resource.create(Attributes.of(ResourceAttributes.SERVICE_NAME, serviceName));
+  }
+
+  @Override
+  public boolean shouldApply(ConfigProperties config, Resource existing) {
+    return ServiceNameChecker.serviceNameNotConfigured(config, existing);
+  }
+
+  @Nullable
+  private Path getJarPathFromProcessHandle() {
+    String[] javaArgs = getProcessHandleArguments.get();
+    for (int i = 0; i < javaArgs.length; ++i) {
+      if ("-jar".equals(javaArgs[i]) && (i < javaArgs.length - 1)) {
+        return Paths.get(javaArgs[i + 1]);
+      }
+    }
+    return null;
+  }
+
+  @SuppressWarnings("unchecked")
+  private static String[] getArgumentsFromProcessHandle() {
```

Comment:
To avoid reflection we could introduce a new module. I guess wether it is worth it depends on whether we'll have more code that need access to jdk9 apis.

Reply:
We could use MR jar too, I guess. I went with that because it seemed the simplest way to call `ProcessHandle` -- I agree that we should replace it with something better when we introduce any more java9+ API usages.